apps/src/lib/metrics/MetricsReporter.ts

```diff
@@ -0,0 +1,104 @@
+import MetricsApi from './MetricsApi';
+
+const isDevelopmentEnvironment =
+  require('../../utils').isDevelopmentEnvironment;
+
+const CHECK_CAN_REPORT_INTERVAL = 30 * 60 * 1000; // 30 minutes in ms
+const LOCAL_STORAGE_KEY_NAME = 'cdo-metrics-reporter-last-check-time';
+
+type LogLevel = 'INFO' | 'WARNING' | 'SEVERE';
+
+/**
+ * Reports logs and metrics, intended primarily for developer-facing
+ * error reporting, metric reporting, and logging.
+ *
+ * For tracking user interactions and behaviors (product-facing),
+ * see {@link AnalyticsReporter} which reports to Amplitude.
+ *
+ * For legacy client-side reporting see {@link firehose} for AWS
+ * Firehose reporting and {@link logToCloud} for New Relic reporting.
+ */
+class MetricsReporter {
+  private lastCheckCanReportTime: number;
+
+  constructor(private readonly metricsApi: MetricsApi) {
+    this.metricsApi = metricsApi;
+    this.lastCheckCanReportTime =
+      parseInt(localStorage.getItem(LOCAL_STORAGE_KEY_NAME) || '0') || 0;
+  }
+
+  logInfo(message: string | object) {
+    this.log('INFO', message);
+    if (isDevelopmentEnvironment()) {
+      console.log('[MetricsReporter] ' + JSON.stringify(message));
+    }
+  }
+
+  logWarning(message: string | object) {
+    this.log('WARNING', message);
+    if (isDevelopmentEnvironment()) {
+      console.warn('[MetricsReporter] ' + JSON.stringify(message));
+    }
+  }
+
+  logError(message: string | object) {
+    this.log('SEVERE', message);
+    if (isDevelopmentEnvironment()) {
+      console.error('[MetricsReporter] ' + JSON.stringify(message));
+    }
+  }
+
+  private log(level: LogLevel, message: string | object) {
+    const payload = {
+      level,
+      message,
+      deviceInfo: this.getDeviceInfo()
+    };
+
+    if (!this.isReportingEnabled()) {
+      this.fallbackLog(payload);
+      return;
+    }
+
+    this.metricsApi.sendLogs([payload]).then(response => {
+      if (!response.ok) {
+        this.fallbackLog(payload);
+      }
+
+      if (response.status === 401) {
+        // Unauthorized response from server; client logging is likely disabled.
+        // We will check again after a time period of CHECK_CAN_REPORT_INTERVAL
+        this.setReportingDisabled();
+      }
+    });
+  }
+
+  private getDeviceInfo(): object {
+    return {
+      user_agent: window.navigator.userAgent,
+      window_width: window.innerWidth,
+      window_height: window.innerHeight,
+      hostname: window.location.hostname,
+      full_path: window.location.href
+    };
+  }
+
+  private fallbackLog(payload: object) {
+    console.log(
```

Comment:
Yeah, this would just be for the experiment/testing phase. This should also only happen if we get an explicit 401 unauthorized from the server. I think long term, yeah we'd just try to send logs as a best effort.